File: Python/12sumList.py


# iterative
# n = number of nodes
# Time: O(n)
# Space: O(1)

def sum_list(head):
  total_sum = 0
  current = head
  while current is not None:
    total_sum += current.val
    current = current.next
  return total_sum

# -----------------------------------

# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)

def sum_list(head):
  if head is None:
    return 0
  return head.val + sum_list(head.next)


# ---------------------------------------
# my implemation
# n = number of nodes
# Time: O(n)
# Space: O(n)

def sum_list(head, sum=0):
  if head is None:
    return sum
  return sum_list(head.next, sum + head.val)

File: Javascript/12sumList.js


// ITERATIVE
// n = # of nodes
// time:  O(n)
// space: O(1)
const sumList = (head) => {
  // todo
  let sum = 0;
  let current = head;
  
  while (current) {
    sum += current.val;
    current = current.next;
  }
  return sum;
};


// RECURSIVE - MY IMPLEMENTATION
// n = # of nodes
// time:  O(n)
// space: O(n)
const sumList = (head, sum=0) => {
  // todo
  if (!head) return sum;
  return sumList(head.next, sum+head.val);
};




// RECURSIVE - STRUCTY - yet another implementation
// n = # of nodes
// time:  O(n)
// space: O(n)
const sumList = (head) => {
  // todo
  if (!head) return 0;
  return head.val + sumList(head.next);
};